tests/integration/pilot/common/routing.go

```diff
@@ -3481,71 +3483,94 @@ spec:
   - "{{$ip}}"
 {{ end }}
   resolution: STATIC
-  endpoints: []
+  endpoints:
+  - address: "10.0.0.1"
   ports:
   - number: 80
     name: http
     protocol: HTTP
 `, map[string]any{"IPs": ips})
 	}
-	ipv4 := "1.2.3.4"
-	ipv6 := "1234:1234:1234::1234:1234:1234"
+	ipv4 := []string{"1.2.3.4", "1.2.3.5"}
+	ipv6 := []string{"1234:1234:1234::1234:1234:1234", "1235:1235:1235::1235:1235:1235"}
 	dummyLocalhostServer := "127.0.0.1"
-	cases := []struct {
-		name string
-		// TODO(https://github.com/istio/istio/issues/30282) support multiple vips
-		ips      string
-		protocol string
-		server   string
-		skipCNI  bool
-		expected []string
-	}{
-		{
-			name:     "tcp ipv4",
-			ips:      ipv4,
-			expected: []string{ipv4},
-			protocol: "tcp",
-		},
-		{
-			name:     "udp ipv4",
-			ips:      ipv4,
-			expected: []string{ipv4},
-			protocol: "udp",
-		},
-		{
-			name:     "tcp ipv6",
-			ips:      ipv6,
-			expected: []string{ipv6},
-			protocol: "tcp",
-		},
-		{
-			name:     "udp ipv6",
-			ips:      ipv6,
-			expected: []string{ipv6},
-			protocol: "udp",
-		},
-		{
-			// We should only capture traffic to servers in /etc/resolv.conf nameservers
-			// This checks we do not capture traffic to other servers.
-			// This is important for cases like app -> istio dns server -> dnsmasq -> upstream
-			// If we captured all DNS traffic, we would loop dnsmasq traffic back to our server.
-			name:     "tcp localhost server",
-			ips:      ipv4,
-			expected: nil,
-			protocol: "tcp",
-			skipCNI:  true,
-			server:   dummyLocalhostServer,
-		},
-		{
-			name:     "udp localhost server",
-			ips:      ipv4,
-			expected: nil,
-			protocol: "udp",
-			skipCNI:  true,
-			server:   dummyLocalhostServer,
-		},
-	}
+
 	for _, client := range flatten(t.Apps.VM, t.Apps.A, t.Apps.Tproxy) {
+		v4, v6 := getSupportedIPFamilies(t, client)
+		var expectedIPv4 []string
+		var expectedIPv6 []string
+		if v4 && v6 {
+			expectedIPv4 = []string{"1.2.3.4", "1.2.3.5"}
+			expectedIPv6 = []string{"1234:1234:1234::1234:1234:1234", "1235:1235:1235::1235:1235:1235"}
+		} else if v4 {
+			expectedIPv4 = []string{"1.2.3.4", "1.2.3.5"}
+			expectedIPv6 = []string{"1234:1234:1234::1234:1234:1234"}
+		} else {
+			expectedIPv4 = []string{"1.2.3.4"}
+			expectedIPv6 = []string{"1234:1234:1234::1234:1234:1234", "1235:1235:1235::1235:1235:1235"}
+		}
```

Comment:
Why is this the expected behavior? It seems a bit un-intuitive that when we are in "v4 only mode" we still return IPv6 addresses, just not all of them?

It may be correct, just not clear why -- a comment would help

Reply:
Right, it is not intuitive. These semantics comes from `GetAllAddressesForProxy`. That function tries to return all available addresses if ambient or dual-stack mode is enabled. Otherwise, it returns all of the addresses for the supported IP family. If there is no IP for the supported family, it fallbacks to the default (first) address.

DNS proxy uses `GetAllAddressesForProxy` to build name table, so it has the same semantics, i.e. in IPv4-only stack it will return all IPv4 addresses for A query and first IPv6 address for AAAA query.

It may be correct - to be honest, I am not sure if this is correct, but this is backward compatible.

Reply:
TBH this behavior seems totally broken. Can we open an issue to track it?

For now doesn't need to block the PR though, good to test our behavior even if its wrong

Reply:
> Can we open an issue to track it?

Of course, I will do it.